File: apps/web/src/lib/questions.ts
const API_URL = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";

export type QuestionAnswer = {
  question: string;
  answer: string;
  confidence: number;
  insufficient_evidence: boolean;
  citations: Array<{
    citation_id: string;
    excerpt: string;
    external_id?: string | null;
    score: number;
    signal: string;
  }>;
  claims: Array<{
    text: string;
    confidence: number;
    citation_ids: string[];
  }>;
  warnings: string[];
  conflicts: string[];
  model?: string | null;
};

export async function askQuestion(
  question: string,
  mode: "hybrid" | "keyword" | "semantic" = "hybrid",
): Promise<QuestionAnswer | null> {
  const response = await fetch(`${API_URL}/api/v1/questions`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ question, mode }),
  });
  if (!response.ok) return null;
  const data = await response.json();
  return {
    ...data,
    warnings: data.warnings ?? [],
    conflicts: data.conflicts ?? [],
  };
}

File: apps/api/src/api/routes/questions.py
from domain.questions import QuestionRequest
from fastapi import APIRouter, Depends

from api.dependencies import RequestServices, get_services
from api.schemas.questions import (
    AnswerClaimResponse,
    CitationResponse,
    QuestionAnswerResponse,
    QuestionBody,
    RetrievalStepResponse,
)

router = APIRouter(tags=["questions"])


@router.post("/questions", response_model=QuestionAnswerResponse)
async def ask_question(
    body: QuestionBody,
    services: RequestServices = Depends(get_services),
) -> QuestionAnswerResponse:
    answer = await services.qa.ask(
        services.owner,
        QuestionRequest(question=body.question, mode=body.mode, limit=body.limit),
        correlation_id=services.correlation_id,
    )
    return QuestionAnswerResponse(
        question=answer.question,
        answer=answer.answer,
        confidence=answer.confidence,
        insufficient_evidence=answer.insufficient_evidence,
        citations=[
            CitationResponse(
                citation_id=item.citation_id,
                chunk_id=item.chunk_id,
                source_id=item.source_id,
                external_id=item.external_id,
                excerpt=item.excerpt,
                score=item.score,
                signal=item.signal.value,
            )
            for item in answer.citations
        ],
        claims=[
            AnswerClaimResponse(
                text=claim.text,
                confidence=claim.confidence,
                citation_ids=claim.citation_ids,
            )
            for claim in answer.claims
        ],
        related_entity_ids=answer.related_entity_ids,
        retrieval_plan=[
            RetrievalStepResponse(
                signal=step.signal.value,
                description=step.description,
                result_count=step.result_count,
            )
            for step in answer.retrieval_plan
        ],
        warnings=answer.warnings,
        conflicts=answer.conflicts,
        model=answer.model,
        created_at=answer.created_at,
    )
